### PR TITLE
[O11y][Mongodb Atlas] Project Logs Datastream

### PR DESCRIPTION
**Pull request by milan-elastic**
Create the package for Mongodb Atlas Integration and Provide the support of **Project logs**.

# Integration release checklist

This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

### All changes

- [x] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/main/CONTRIBUTING.md)
- [ ] Supported versions of the monitoring target are documented
- [ ] Supported operating systems are documented (if applicable)
- [x] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist
- [x] Documentation exists, [useful guidelines](https://github.com/elastic/integrations/blob/main/docs/documentation_guidelines.md) to follow
- [x] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
- [x] At least a manual test with ES / Kibana / Agent has been performed.
- [x] Required Kibana version set to: 8.12.0
- [ ] TSDB enablement for the data stream

### Data dataset changes

_This entry is currently only recommended. It will be mandatory once we provide better support for it._

- [x] Sample event (`sample_event.json`) exists

### Related Issue
- Relates https://github.com/elastic/obs-infraobs-team/issues/1376

### Screenshots

mongodb_atlas-project-dashboard

### Known Issues:
Currently Following challenges are still open to solve for this PR:
- https://github.com/elastic/obs-infraobs-team/issues/1350
- https://github.com/elastic/obs-infraobs-team/issues/1351
- https://github.com/elastic/obs-infraobs-team/issues/1372

### CODE REVIEW

**Review by niraj-elastic — APPROVED** [review]
LGTM

File: packages/mongodb_atlas/data_stream/project/agent/stream/input.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        state.want_more ? 
+          state 
+        :
+          state.with({
+            "minDate": state.?cursor.last_timestamp.orValue(now - duration("30m")),
+            "maxDate": now
+        })
+      ).as(state, 
+        request(
+          "GET",
```
Comment (ishleenk17) [inline packages/mongodb_atlas/data_stream/project/agent/stream/input.yml.hbs:47]: Is the script completely similar to the previous datastreams?

File: packages/mongodb_atlas/data_stream/project/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        multi: false
+        required: true
+        show_user: false
+      - name: http_client_timeout
+        type: text
+        title: HTTP Client Timeout
+        description: Duration before declaring that the HTTP client connection has timed out. Valid time units are h (hours), m (minutes) and s (seconds). If the data collection period is extensive, ensure that this parameter's value is appropriately long.
+        multi: false
+        required: true
+        show_user: false
```
Comment (shmsr) [inline packages/mongodb_atlas/data_stream/project/manifest.yml:31]:
```suggestion
        required: true
        show_user: false
```

It is a required field but not shown to the user. I think `required: false` is the correct one? You already have the default value.

File: packages/mongodb_atlas/data_stream/project/agent/stream/input.yml.hbs (hunk 1 above)
Comment (milan-elastic) [reply packages/mongodb_atlas/data_stream/project/agent/stream/input.yml.hbs:47]: Earlier, for MongoDB_audit and MongoDB_database, we needed to make multiple API calls to collect data. However, for Organization and Project, the API call is straightforward. Therefore, the data collection script differs from the previous datastreams(MongoDB_audit and MongoDB_database).

**Comment by shmsr** [comment]
Review:

1. Script changes are not mandatory. LOCALE.ROOT improvement I got it from [these](https://www.elastic.co/guide/en/elasticsearch/painless/7.2/painless-walkthrough.html#:~:text=player%E2%80%99s%20last%20names-,upper%20case,-%3A) examples.
2. Fixed the field descriptions

```diff
diff --git a/packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml b/packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml
index 7be93e8dd6..4cb3a33801 100644
--- a/packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml
+++ b/packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml
@@ -5,10 +5,10 @@ processors:
       if: ctx.response == "The application encountered an error while retrieving the current page"
       tag: drop_retry_page_events
   - fingerprint:
-      fields: 
+      fields:
         - response.id
       target_field: _id
-      ignore_missing: true 
+      ignore_missing: true
   - set:
       field: ecs.version
       value: 8.11.0
@@ -109,15 +109,18 @@ processors:
             if (v instanceof Map) {
               v = keysToSnakeCase(v);
             } else if (v instanceof List) {
-              for (int i = 0; i < v.size(); i++) {
-                def item = v.get(i);
+              def newList = new ArrayList();
+              for (item in v) {
                 if (item instanceof Map) {
-                  v.set(i, keysToSnakeCase(item));
+                  newList.add(keysToSnakeCase(item));
+                } else {
+                    newList.add(item);
                 }
               }
+              v = newList;
             }
 
-            k = regex.matcher(k).replaceAll('$1_$2').toLowerCase();
+            k = regex.matcher(k).replaceAll('$1_$2').toLowerCase(Locale.ROOT);
             snakeCaseMap.put(k, v);
           }
           return snakeCaseMap;
@@ -327,4 +330,4 @@ on_failure:
   - append:
       field: event.kind
       value: pipeline_error
-      allow_duplicates: false      
+      allow_duplicates: false
diff --git a/packages/mongodb_atlas/data_stream/project/fields/fields.yml b/packages/mongodb_atlas/data_stream/project/fields/fields.yml
index 9373fb25b4..fa950840bd 100644
--- a/packages/mongodb_atlas/data_stream/project/fields/fields.yml
+++ b/packages/mongodb_atlas/data_stream/project/fields/fields.yml
@@ -84,8 +84,7 @@
               description: Name of the measurement whose value went outside the threshold.
             - name: unit
               type: keyword
-              description: |-
-                Relevant units for value.
+              description: Unit for the value.
             - name: value
               type: float
               description: Value of the metric.
@@ -109,7 +108,7 @@
                   description: Name of the stream processing instance associated with the event.
             - name: name
               type: keyword
-              description: Error message linked to the stream processor associated with the event.
+              description: Name of the stream processor associated with the event.
             - name: state
               type: keyword
               description: State of the stream processor associated with the event.
diff --git a/packages/mongodb_atlas/data_stream/project/manifest.yml b/packages/mongodb_atlas/data_stream/project/manifest.yml
index c82f8176f6..562182ee56 100644
--- a/packages/mongodb_atlas/data_stream/project/manifest.yml
+++ b/packages/mongodb_atlas/data_stream/project/manifest.yml
@@ -10,7 +10,7 @@ streams:
       - name: groupId
         type: text
         title: Group ID
-        description: Identifier string that identifies your project. Groups and projects are synonymous terms. Your group ID is the same as your project ID. Ex. 32b6e34b3d91647abb20e7b8
+        description: Identifier string that identifies your project. Groups and projects are synonymous; your group ID is the same as your project ID (e.g., 32b6e34b3d91647abb20e7b8).
         multi: false
         required: true
         show_user: true
@@ -36,7 +36,7 @@ streams:
         multi: false
         required: false
         show_user: false
-        description: The request tracer logs requests and responses to the agent's local file-system for debugging configurations. Enabling this request tracing compromises security and should only be used for debugging. See [documentation](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_filename) for details.
+        description: The request tracer logs requests and responses to the agent's local file-system for debugging configurations. This feature should only be used for debugging as it may compromise security. See [documentation](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_filename) for details.
       - name: tags
         type: text
         title: Tags
@@ -60,4 +60,4 @@ streams:
         required: false
         show_user: false
         description: >-
-          Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This executes in the agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/fleet/current/elastic-agent-processor-configuration.html) for details.
+          Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This is executed in the Agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/fleet/current/elastic-agent-processor-configuration.html) for details.
```

File: packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+      value: event
+  - set:
+      field: event.module
+      value: mongodb_atlas
+  - set:
+      field: event.category
+      value: ["configuration", "database"]
+  - set:
+      field: event.type
+      value: ["info", "access", "change"]
```
Comment (shmsr) [inline packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml:26]: For event.type, I think the categorization could be better:

- set "info" always
- append "access" if and only if (iff) `ctx.mongodb_atlas?.project?.operation?.type == 'READ'`
- append "change" iff `ctx.mongodb_atlas?.project?.operation?.type` contains create/update/delete/insert etc. (see [this](https://github.com/mongodb/mongo/blob/79e6457fbfb079973fdaa25214c04ec909668496/src/mongo/shell/bench.h#L67))
- append "error" iff `ctx.error?.message != null || ctx.mongodb_atlas?.project?.processor?.error_msg != null`


Currently, event.type is set to 3 values irrespective of the type of event received.

File: packages/mongodb_atlas/data_stream/project/manifest.yml (hunk 2 above)
Comment (milan-elastic) [reply packages/mongodb_atlas/data_stream/project/manifest.yml]: Addressed!!

File: packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+  - append:
+      field: event.type
+      value: access
+      allow_duplicates: false
+      if: ctx.response?.opType != null && ctx.response?.opType == 'read'
+  - append:
+      field: event.type
+      value: change
+      allow_duplicates: false
+      if: ctx.response?.opType != null && ['create', 'update', 'delete', 'insert'].contains(ctx.response.opType)
```
Comment (shmsr) [inline packages/mongodb_atlas/data_stream/project/elasticsearch/ingest_pipeline/default.yml:36]: Are we sure the values of opType always returns lowercase keywords like 'create', 'update', 'delete' or 'insert'?

If not, there's convert the `ctx.response.opType` to lowercase using `toLowerCase()` and do the contains check.

**Review by shmsr — APPROVED** [review]
Left one comment. Please address and then we are ready to merge.

**Comment by ishleenk17** [comment]
Folks, are we good here ?
@milan-elastic @shmsr

**Review by ishleenk17 — APPROVED** [review]
Looks good!

**Pull request merged**